### PR TITLE
Localization using pinger bearing

### PR DESCRIPTION
**Pull request by irwineffect**
Modified the localization system to use the pinger bearing instead of position estimate from old hydrophone trilateration. Not entirely sure if I did it right, would love some review from @BigDCross.

---
This change is [Reviewable](https://reviewable.io/reviews/palouserobosub/robosub/307)

### CODE REVIEW

**Comment by skallaher** [comment]
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[param/cobalt.yaml, line 161 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kveds9__LzE3XFMnZ24:-Kveds9__LzE3XFMnZ25:b-1k410s) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/param/cobalt.yaml#L161)):*
> ```YAML
>         z: 0.015
>     pinger:
>         depth: -4.1
> ```

Is there a reason for this to be in `cobalt.yaml` rather than `localization.yaml`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by irwineffect** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[param/cobalt.yaml, line 161 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kveds9__LzE3XFMnZ24:-KveqZrGvCIBDNgRfkLd:b-h1ac5q) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/param/cobalt.yaml#L161)):*
<details><summary><i>Previously, skallaher (Sean Kallaher) wrote…</i></summary><blockquote>

Is there a reason for this to be in `cobalt.yaml` rather than `localization.yaml`?
</blockquote></details>

I was debating the same thing. If you think it should go in `localization.yaml`, I would totally agree.

---

*[src/localization/particle_filter.cpp, line 257 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kver7US5jRjSIuN7VnF:-Kver7US5jRjSIuN7VnG:bkmlt16) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/src/localization/particle_filter.cpp#L257)):*
> ```cpp
>     obs(0, 0) = hx/magnitude;
>     obs(1, 0) = hy/magnitude;
>     obs(2, 0) = hz/magnitude;
> ```

I just realized, shouldn't this vector need to be rotated based on the sub's orientation?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by skallaher** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[param/cobalt.yaml, line 161 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kveds9__LzE3XFMnZ24:-KverW4TMVC30Vvh9zNJ:b-dheekb) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/param/cobalt.yaml#L161)):*
<details><summary><i>Previously, irwineffect wrote…</i></summary><blockquote>

I was debating the same thing. If you think it should go in `localization.yaml`, I would totally agree.
</blockquote></details>

I think it should be. At least for now since no other system uses it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by bkallaher** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/localization/particle_filter.cpp, line 257 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kver7US5jRjSIuN7VnF:-Kvf8-NJv09i7NbMZ5mq:b50fa54) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/src/localization/particle_filter.cpp#L257)):*
<details><summary><i>Previously, irwineffect wrote…</i></summary><blockquote>

I just realized, shouldn't this vector need to be rotated based on the sub's orientation?
</blockquote></details>

Probably not as your bearing is already relative to the course of the sub. So unless you want to have an absolute bearing it should be fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by irwineffect** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/localization/particle_filter.cpp, line 257 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kver7US5jRjSIuN7VnF:-KvfAkEfpxzoDE2G_sw9:brntziy) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/src/localization/particle_filter.cpp#L257)):*
<details><summary><i>Previously, bkallaher (Brandon Kallaher) wrote…</i></summary><blockquote>

Probably not as your bearing is already relative to the course of the sub. So unless you want to have an absolute bearing it should be fine.
</blockquote></details>

This part of the code is faking out what sort of bearing reading each particle would observe given it's position, which is then compared to the incoming bearing reading from the hydrophones. I think as the code is right now it calculates the bearing as if the particles had zero RPY, which then needs to be transformed giving the sub's actual orientation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by irwineffect** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/localization/particle_filter.cpp, line 257 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kver7US5jRjSIuN7VnF:-KvfCCp6UwyhwRhfO-qc:bci3ru9) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/src/localization/particle_filter.cpp#L257)):*
<details><summary><i>Previously, irwineffect wrote…</i></summary><blockquote>

This part of the code is faking out what sort of bearing reading each particle would observe given it's position, which is then compared to the incoming bearing reading from the hydrophones. I think as the code is right now it calculates the bearing as if the particles had zero RPY, which then needs to be transformed giving the sub's actual orientation.
</blockquote></details>

In other words, what I'm saying is that the hydrophone bearing is relative to the sub's coordinate frame, while right now these particle's simulated bearings are in the global frame, and then the code directly compares the two. One or the other needs to be translated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by skallaher** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/localization/particle_filter.cpp, line 257 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kver7US5jRjSIuN7VnF:-KverhhEpgw4T2VSq9_a:b-wymn77) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/src/localization/particle_filter.cpp#L257)):*
<details><summary><i>Previously, irwineffect wrote…</i></summary><blockquote>

In other words, what I'm saying is that the hydrophone bearing is relative to the sub's coordinate frame, while right now these particle's simulated bearings are in the global frame, and then the code directly compares the two. One or the other needs to be translated.
</blockquote></details>

Would it not be easier to translate just the one bearing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by Sindorman** [comment]
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by irwineffect** [comment]
Review status: 7 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/localization/particle_filter.cpp, line 257 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kver7US5jRjSIuN7VnF:-Kviiy0WUgNmFeQyS1A0:b-5cjxxm) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/src/localization/particle_filter.cpp#L257)):*
<details><summary><i>Previously, skallaher (Sean Kallaher) wrote…</i></summary><blockquote>

Would it not be easier to translate just the one bearing?
</blockquote></details>

Yeah, I could translate the actual pinger reading (and it would be more computationally efficient), but IMHO it doesn't make as much physical sense to do that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by irwineffect** [comment]
Review status: 5 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[param/cobalt.yaml, line 161 at r1](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-Kveds9__LzE3XFMnZ24:-KvikIHeMK8RlFwj_p-v:b-896fix) ([raw file](https://github.com/palouserobosub/robosub/blob/a9de7eeca2f4307694838b8a8e3de9bec0150147/param/cobalt.yaml#L161)):*
<details><summary><i>Previously, skallaher (Sean Kallaher) wrote…</i></summary><blockquote>

I think it should be. At least for now since no other system uses it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307)*

**Comment by skallaher** [comment]
:lgtm: Will need to evaluate some numbers for accuracy and such in future.

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-:-Kvik_uHiAR97HRLw-a9:b-as8tzv)*

**Comment by skallaher** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-:-Kvk9YzqQZQB6AFm8E_z:bnfp4nl)*

**Comment by connor-wool** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palouserobosub/robosub/307#-:-KwMCzu4AxRu2EYir4Lg:bnfp4nl)*
